core/common/src/main/java/com/google/jetpackcamera/core/common/RefCounted.kt

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jetpackcamera.core.common
+
+import android.util.Log
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+
+/**
+ * A thread-safe, lock-free class for managing the lifecycle of an object using ref-counting.
+ *
+ * The object that is ref-counted can be initialized late using [initialize].
+ *
+ * @param[debugRefCounts] whether to print debug log statements.
+ * @param[onRelease] a block that will be invoked once when the ref-count reaches 0.
+ */
+class RefCounted<T : Any>(
+    private val debugRefCounts: Boolean = false,
+    private val onRelease: (T) -> Unit = {}
```

Comment:
From an API perspective, I would force this to be a required parameter. After all, Kotlin doesn't really need reference counting if onRelease is a no-op. People coming from other languages may try to use RefCouted where `shared_ptr` or `sp` would typically be used (even though it is not needed).

Reply:
Yeah that's probably better- changed it to be required.

The only case I can think of where you would want a no-op `onRelease` is for flow control. That is, wrap the object and perform tasks only if you can `acquire()`. However, there are probably better ways to do that, and you can always pass in a no-op `onRelease` if that's what you want.